kernel/src/collections/list_v.rs

```diff
@@ -0,0 +1,606 @@
+                    ghost_state.borrow_mut().points_to_map.tracked_remove(1);
+                    ghost_state.borrow_mut().points_to_map.tracked_map_keys_in_place(
+                        Map::<nat, nat>::new(
+                            |j: nat| 1 <= j && j < ghost_state@.cells.len(),
+                            |j: nat| (j + 1) as nat,
+                        ),
+                    );
+                    ghost_state.borrow_mut().points_to_map.tracked_insert(0, head_points_to);
+                    assert forall|i: nat|
+                        1 <= i < ghost_state@.cells.len() implies well_formed_node(
+                        ghost_state,
+                        i,
+                    ) by {
+                        assert(old(ghost_state)@.cells[(i + 1) as int]
+                            == ghost_state@.cells[i as int]);
+                        assert(old(ghost_state)@.points_to_map.dom().contains((i + 1) as nat)
+                            == ghost_state@.points_to_map.dom().contains(i));
+                        assert(old(ghost_state)@.points_to_map[(i + 1) as nat]
+                            == ghost_state@.points_to_map[i]);
+                    }
+                }
+            },
+            None => {
+                proof {
+                    assert(ghost_state@.cells.len() == 1);
+                }
+
+                self.head.0.write(Tracked(&mut head_points_to), None);
+
+                proof {
+                    ghost_state.borrow_mut().points_to_map.tracked_insert(0, head_points_to);
+                }
+            },
+        }
+        remove
+    }
+
+    pub fn iter(&'a self, ghost_state: &Tracked<GhostState<'a, T>>) -> (res: ListIteratorV<'a, T>)
+        requires
+            ghost_state@.cells.len() >= 1,
+            forall|i: nat|
+                0 <= i < ghost_state@.cells.len()
+                    ==> #[trigger] ghost_state@.points_to_map.dom().contains(i)
+                    && ghost_state@.points_to_map[i].is_init() && ghost_state@.points_to_map[i].id()
+                    == ghost_state@.cells[i as int].id(),
+            forall|i: nat|
+                0 <= i < (ghost_state@.cells.len() - 1) as nat
+                    ==> match #[trigger] ghost_state@.points_to_map[i].value() {
+                    Option::Some(_) => true,
+                    Option::None => false,
+                },
+            match ghost_state@.points_to_map[(ghost_state@.cells.len() - 1) as nat].value() {
+                Option::Some(_) => false,
+                Option::None => true,
+            },
+            ghost_state@.cells[0].id() == self.head.0.id(),
+        ensures
+            res.cur == ghost_state@.points_to_map[0].value(),
+            res.index@ == 0,
+    {
+        proof {
+            assert(ghost_state@.points_to_map.dom().contains(0));
+        }
+        ListIteratorV::<T>::new(self, ghost_state)
+    }
+}
+
+fn main() {
```

Comment:
Why do we have a `main`?

Reply:
Verus requires so.

Reply:
Are you sure it is the case if you verify `list_v.rs` as part of the kernel library? In this case it should not have to have a main

Reply:
Ok, but if you verify this file individually, Verus will throw error if there is not a main function. So why don't we write one here? It should not affect code tidiness much.